lib/server-app.js

```diff
@@ -161,7 +161,28 @@ proto.middleware = function(name, paths, handler) {
 
   this._skipLayerSorting = true;
   this.use(paths, handler);
-  this._router.stack[this._router.stack.length - 1].phase = fullPhaseName;
+
+  // Other handlers can be added to the stack, for example,
+  // NewRelic adds sentinel handler. We need to search the stack
+  var found = false;
+  for (var k = this._router.stack.length - 1; k >= 0 && !found; k--) {
+    if (this._router.stack[k].handle === handler ||
+      // NewRelic replaces the handle and keeps it as __NR_original
+      this._router.stack[k].handle['__NR_original'] === handler
```

Comment:
Is this optimisation worth the extra code? We will be checking most of the middleware entries using the full property scan anyways.